File: backend/app/models/personal_recipe.py
"""Personal recipe models for the personal recipe management feature.

This module contains Pydantic models for creating, updating, and responding
to personal recipe operations.
"""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class IngredientCreate(BaseModel):
    """Request model for creating an ingredient."""

    name: str = Field(..., min_length=1, max_length=255)
    quantity: float = Field(..., gt=0)
    unit: str = Field(..., max_length=50)
    note: str | None = Field(None, max_length=255)


class InstructionStepCreate(BaseModel):
    """Request model for creating an instruction step."""

    instruction: str = Field(..., min_length=1)


class PersonalRecipeCreate(BaseModel):
    """Request model for creating a personal recipe."""

    title: str = Field(..., min_length=1, max_length=255)
    image_url: str | None = Field(None, max_length=500)
    servings: int = Field(..., ge=1, le=20)
    prep_time_minutes: int | None = Field(None, ge=0)
    cook_time_minutes: int | None = Field(None, ge=0)
    tags: list[str] | None = None
    ingredients: list[IngredientCreate] = Field(..., min_length=1)
    steps: list[InstructionStepCreate] = Field(..., min_length=1)


class PersonalRecipeUpdate(BaseModel):
    """Request model for updating a personal recipe.

    All fields are optional for partial updates. However, if ingredients or
    steps are provided, they must have at least one item.
    """

    title: str | None = Field(None, min_length=1, max_length=255)
    image_url: str | None = Field(None, max_length=500)
    servings: int | None = Field(None, ge=1, le=20)
    prep_time_minutes: int | None = Field(None, ge=0)
    cook_time_minutes: int | None = Field(None, ge=0)
    tags: list[str] | None = None
    ingredients: list[IngredientCreate] | None = Field(None, min_length=1)
    steps: list[InstructionStepCreate] | None = Field(None, min_length=1)


class IngredientResponse(BaseModel):
    """Response model for an ingredient."""

    id: str
    name: str
    quantity: float
    unit: str
    note: str | None
    sort_order: int


class InstructionStepResponse(BaseModel):
    """Response model for an instruction step."""

    id: str
    step_number: int
    instruction: str


class PersonalRecipeResponse(BaseModel):
    """Response model for a full personal recipe with nested data."""

    id: str
    user_id: str
    title: str
    image_url: str | None
    servings: int
    prep_time_minutes: int | None
    cook_time_minutes: int | None
    tags: list[str] | None
    source_recipe_id: str | None
    created_at: datetime
    updated_at: datetime
    ingredients: list[IngredientResponse]
    steps: list[InstructionStepResponse]


class PersonalRecipeSummary(BaseModel):
    """Summary model for search results.

    Used in unified search to identify the source of the recipe.
    """

    id: str
    title: str
    image_url: str | None
    source: Literal["personal", "api"]


class PersonalRecipeListResponse(BaseModel):
    """Response model for list of personal recipes."""

    recipes: list[PersonalRecipeSummary]
    count: int

File: mobile/types/personal-recipe.ts
/**
 * Personal Recipe Types
 *
 * TypeScript types for user's personal recipes.
 */

/** Ingredient for personal recipe creation/update */
export interface IngredientInput {
  name: string;
  quantity: number;
  unit: string;
  note?: string | null;
}

/** Ingredient response from API */
export interface IngredientResponse {
  id: string;
  name: string;
  quantity: number;
  unit: string;
  note: string | null;
  sort_order: number;
}

/** Instruction step for creation/update */
export interface InstructionStepInput {
  instruction: string;
}

/** Instruction step response from API */
export interface InstructionStepResponse {
  id: string;
  step_number: number;
  instruction: string;
}

/** Data for creating a new personal recipe */
export interface PersonalRecipeCreate {
  title: string;
  image_url?: string | null;
  servings: number;
  prep_time_minutes?: number | null;
  cook_time_minutes?: number | null;
  tags?: string[] | null;
  ingredients: IngredientInput[];
  steps: InstructionStepInput[];
}

/** Data for updating a personal recipe (all fields optional) */
export interface PersonalRecipeUpdate {
  title?: string;
  image_url?: string | null;
  servings?: number;
  prep_time_minutes?: number | null;
  cook_time_minutes?: number | null;
  tags?: string[] | null;
  ingredients?: IngredientInput[];
  steps?: InstructionStepInput[];
}

/** Full personal recipe response */
export interface PersonalRecipeResponse {
  id: string;
  user_id: string;
  title: string;
  image_url: string | null;
  servings: number;
  prep_time_minutes: number | null;
  cook_time_minutes: number | null;
  tags: string[] | null;
  source_recipe_id: string | null;
  created_at: string;
  updated_at: string;
  ingredients: IngredientResponse[];
  steps: InstructionStepResponse[];
}

/** Summary for list views */
export interface PersonalRecipeSummary {
  id: string;
  title: string;
  image_url: string | null;
  source: 'personal';
}

/** List response from API */
export interface PersonalRecipeListResponse {
  recipes: PersonalRecipeSummary[];
  count: number;
}

/** Image upload response */
export interface ImageUploadResponse {
  url: string;
  filename: string;
}

/** Unified search result item */
export interface UnifiedRecipeSummary {
  id: string;
  name: string;
  thumbnail: string;
  source: 'personal' | 'api';
}

/** Unified search response */
export interface UnifiedSearchResponse {
  recipes: UnifiedRecipeSummary[];
  personal_count: number;
  api_count: number;
}

/** Common unit options for ingredient input */
export const UNIT_OPTIONS = [
  { label: 'g', value: 'g' },
  { label: 'kg', value: 'kg' },
  { label: 'ml', value: 'ml' },
  { label: 'L', value: 'L' },
  { label: 'c. a cafe', value: 'tsp' },
  { label: 'c. a soupe', value: 'tbsp' },
  { label: 'tasse', value: 'cup' },
  { label: 'piece', value: 'piece' },
  { label: 'pincee', value: 'pinch' },
  { label: 'au gout', value: 'to taste' },
] as const;
